streaming-payments.md

```diff
@@ -0,0 +1,375 @@
+condition = sha256(fulfillment)
+```
+
+To run this new version of the shop, type this into your terminal:
+
+```sh
+npm install ilp-packet
+node ./shop2.js
+```
+
+The ILP paacket is not very useful in a one-transfer payment, but in future tutorials, we see how connectors can forward an
+incoming transfer, and thus connect one ledger to other ledgers. When that happens, we say the whole sender
+to receiver process is the "payment", and each link in the chain is a "transfer", so one payment consists of 
+one or more transfers. The sender can then put information in the ILP packet which the receiver can use
+to generate the fulfillment, and if that information would be tempered with by connectors along the way,
+this would be detected because the fulfillment would not match.
+The ILP Payment Packet is serialized into OER and then Base64-encoded before it's added to the transfer as a Memo.
+Note that this assumes that the ledger supports adding a memo to the transfer, but luckily, most ledgers do.
+
+## Payment tool
+
+The following is mainly a mix between the `pay.js` and `proxy.js` scripts from the Letter Shop tutorial,
+that can pay for content in reaction to an `Pay` header, and then stream that content to the console
+as it comes in. Save it as `tool.js`:
+```js
+const IlpPacket = require('ilp-packet')
+const Plugin = require('ilp-plugin-xrp-escrow')
+const crypto = require('crypto')
+const fetch = require('node-fetch')
+const uuid = require('uuid/v4')
+function base64 (buf) { return buf.toString('base64').replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '') }
+function hash (secret) { return crypto.createHash('sha256').update(secret).digest() }
+function hmac (secret, input) { return crypto.createHmac('sha256', secret).update(input).digest() }
+
+const plugin = new Plugin({
+  secret: 'sndb5JDdyWiHZia9zv44zSr2itRy1',
+  account: 'rGtqDAJNTDMLaNNfq1RVYgPT8onFMj19Aj',
+  server: 'wss://s.altnet.rippletest.net:51233',
+  prefix: 'test.crypto.xrp.'
+})
+
+let counter = 0
+
+function sendTransfer (obj) {
+  obj.id = uuid()
+  obj.from = plugin.getAccount()
+  // to
+  obj.ledger = plugin.getInfo().prefix
+  // amount
+  // executionCondition
+  obj.expiresAt = new Date(new Date().getTime() + 1000000).toISOString()
+  return plugin.sendTransfer(obj).then(function () {
+    return obj.id
+  })
+}
+
+plugin.connect().then(function () {
+  return fetch('http://localhost:8000/')
+}).then(function (inRes) {
+  inRes.body.pipe(process.stdout)
+  const payHeaderParts = inRes.headers.get('Pay').split(' ')
+  console.log(payHeaderParts)
+  // e.g. Pay: 1 test.crypto.xrp.asdfaqefq3f.26wrgevaew SkTcFTZCBKgP6A6QOUVcwWCCgYIP4rJPHlIzreavHdU
+  setInterval(function () {
+    const ilpPacketContents = {
+      account: payHeaderParts[1] + '.' + (++counter),
+      amount: '1',
+      data: ''
+    }
+    const fulfillment = hmac(Buffer.from(payHeaderParts[2], 'base64'), ilpPacketContents.account)
```

Comment:
This is the kind of thing that people shouldn't be implementing themselves. If you ever send a second payment with a condition you used before, a connector can take the money before it gets to the receiver. If the sender is generating the fulfillment they should always use some of their own randomness, rather than relying on someone else. 

This is why I think it's better to use libraries like PSK that prevent you from shooting yourself in the foot in this way. I don't think using PSK here would detract too much from the explanatory value because you would use the function called `createPacketAndCondition` and could explain that that generates those two from the supplied parameters and a nonce such that the receiver will be able to regenerate the same value and check that the packet is intact.

Reply:
OK, I proposed https://github.com/interledger/rfcs/issues/310

Reply:
it uses the code directly from https://interledger.org/rfcs/0016-pre-shared-key/draft-3.html#1-payment-creation now